feat: EVM emulator docs

uF4No opened the pull request:
# Description

EVM Emulator documentation

- Overview page and FAQ in the ZKsync Era section of the docs: ZKsync Era > Unique Features > EVM Bytecode Emulator
- More detailed docs in the Protocol section of the docs with multiple pages: ZKsync protocol > Contracts > EVM Bytecode Emulator

## Linked Issues

N/A

## Additional context

Required for EVM Emulator release which is part of v27

uF4No committed:
feat: base emulator docs

uF4No committed:
feat: emulator protocol docs

uF4No committed:
fix: hardcoded staging link

uF4No committed:
fix: broken links

uF4No committed:
fix: broken links again

uF4No committed:
fix: more broken links

sarahschwartz commented on content/00.zksync-era/30.unique-features/40.evm-emulator/01.evm-emulator.md:
```diff
@@ -0,0 +1,36 @@
+title: Overview
+description: The special contracts that allows executing EVM bytecode on EraVM
+---
+ZKsync Era integrates an EVM bytecode emulator that enables compatibility with EVM bytecode by translating EVM instructions into EraVM
```
Just a suggestion, this sentence feels a bit long.

```suggestion
ZKsync Era integrates an EVM bytecode emulator that enables compatibility with EVM bytecode. The emulator works by translating EVM instructions into EraVM
```

sarahschwartz commented on content/00.zksync-era/30.unique-features/40.evm-emulator/01.evm-emulator.md:
```diff
@@ -0,0 +1,36 @@
+instructions at execution time.
+
+This allows developers to deploy and interact with smart contracts without the need of using the ZKsync custom compilers
+(`zksolc` and `zkvyper`) Additionally, standard EVM tooling like Foundry and Hardhat works out of the box without the need for plugins.
```
```suggestion
(`zksolc` and `zkvyper`). Additionally, standard EVM tooling like Foundry and Hardhat works out of the box without the need for plugins.
```

uF4No committed:
Merge branch 'main' into feat/emulator

sarahschwartz reviewed (commented):
Looks great! I just have a nit. 

And random question, do we know that reentrancy works the same with the emulator? I don't think that's mentioned at all.

uF4No committed:
Apply suggestions from code review

Co-authored-by: Sarah Schwartz <58856580+sarahschwartz@users.noreply.github.com>

uF4No committed:
feat: adds more internal links

0xVolosnikov commented:
> do we know that reentrancy works the same with the emulator

It works in the same way as in EVM

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/_dir.yml:
```diff
@@ -0,0 +1 @@
+title: EVM Bytecode Emulator
```
Imo "EVM Emulation" sounds better, wdyt?

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/01.evm-emulator.md:
```diff
@@ -0,0 +1,41 @@
+
+## Main Features
+
+- **Standard Compiler Support**: Contracts do not need to be compiled with ZKsync-specific compilers (`zksolc` or `zkvyper`).
```
```suggestion
- **EVM Bytecode Support**: Contracts do not need to be compiled with ZKsync-specific compilers (`zksolc` or `zkvyper`).
```

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/01.evm-emulator.md:
```diff
@@ -0,0 +1,41 @@
+- **Address Derivation Consistency**: Addresses generated using `create` and `create2` match the behavior on the EVM.
+- **Pre Deployed Contracts**: Provides contracts such as `create2` , `multicall3`  and
+  `singletonFactory` (ERC2470). Refer to the [protocol documentation](../../../zksync-protocol/contracts/evm-emulator/pre-deployed-contracts) for
+  additional the full list.
```
```suggestion
  the full list.
```

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/01.evm-emulator.md:
```diff
@@ -0,0 +1,41 @@
+
+## Limitations
+
+- **Debugging**: EVM-like debugging is not available due to missing RPC methods.
```
```suggestion
- **Debugging**: EVM-compatible debugging is not supported yet due to EraVM-specific nuances of the system.
```

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/01.evm-emulator.md:
```diff
@@ -0,0 +1,41 @@
+
+- **Debugging**: EVM-like debugging is not available due to missing RPC methods.
+- **Gas Behavior**: Gas calculations follow EraVM logic, not EVM gas logic.
+- **Unsupported Opcodes**: The same opcodes unsupported in EraVM remain unsupported. See the
```
Let's list unsupported opcodes here

0xVolosnikov commented:
I think we need to add some notification that EVM emulator is an **optional** feature and is not yet available on Era

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/02.faq.md:
```diff
@@ -0,0 +1,40 @@
+---
+
+#### Can I use default Foundry instead of Foundry-zksync?
+Yes. Although as gas logic on EraVM differs from the EVM, it's recommended to use the `--skip-simulation` flag.
```
```suggestion
Yes. But for scripts execution (`forge script ...`) it's recommended to use the `--skip-simulation` flag, as gas logic on EraVM differs from the EVM and foundry can estimate gas incorrectly.
```

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/02.faq.md:
```diff
@@ -0,0 +1,40 @@
+The EVM emulator targets **Cancun**.
+
+#### Given that on EraVM contract deployment is done via the `ContractDeployer`, how does deployment work using the EVM Emulator?
+You can deploy contracts just like on Ethereum (e.g., a transaction to `0x0` with the contract bytecode) and use tools like Hardhat
```
```suggestion
You can deploy contracts just like on Ethereum (e.g., a transaction without field `to` and with the contract init code as calldata) and use tools like Hardhat
```

0xVolosnikov commented on content/00.zksync-era/30.unique-features/40.evm-emulator/02.faq.md:
```diff
@@ -0,0 +1,40 @@
+See [Limitations page in the protocol documentation](../../../zksync-protocol/contracts/evm-emulator/limitations).
+
+#### What's the behavior of `block.gaslimit`? Does it return the batch limit or the transaction limit?
+It returns `1125899906842624n`, the batch limit.
```
```suggestion
It returns `1125899906842624n`. In practice this is meaningless value and should not be used.
```

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/00.overview.md:
```diff
@@ -0,0 +1,65 @@
+  Instead of running native EraVM bytecode, the system routes execution through the emulator.
+- **EVM opcodes are interpreted at runtime.** The emulator translates EVM instructions into **EraVM operations**, ensuring execution
+  adheres to EVM semantics as closely as possible.
+- **Gas accounting is based on EraVM (ergs).** While transaction fees remain lower than Layer 1, the cost of executing EVM contracts
```
Everyting related to gas is confusing (because our gas model is confusing). We need to think about it a bit. Maybe we should have a separate page with a clear explanation of how our gas model works

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/00.overview.md:
```diff
@@ -0,0 +1,65 @@
+|----------|------------|-----------|
+| **Remix** | ✅ Works (Deploy, Call, Send TX) | Debugging **not supported**. |
+| **Foundry** | ✅ Works (`forge`, `cast`) | **Recommended:** Use `--skip-simulation` due to gas calculation differences. |
+| **Anvil** | ✅ Works | Forking from a chain with the EVM Emulator works. **To use EraVM-specific features, it's recommended [anvil-zksync instead](../../../zksync-era/tooling/local-setup/anvil-zksync-node)**. |
```
> Forking from a chain with the EVM Emulator works

But if something EraVM-specific is called (e.g. EraVM contract), it will explode almost 100%.

uF4No replied:
I'll add that for zksync specific features anvil-zksync is required instead of recommended.

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/00.overview.md:
```diff
@@ -0,0 +1,65 @@
+| **Tool** | **Support** | **Notes** |
+|----------|------------|-----------|
+| **Remix** | ✅ Works (Deploy, Call, Send TX) | Debugging **not supported**. |
+| **Foundry** | ✅ Works (`forge`, `cast`) | **Recommended:** Use `--skip-simulation` due to gas calculation differences. |
```
```suggestion
| **Foundry** | ✅ Works (`forge`, `cast`) | **Recommended:** For scripts use `--skip-simulation` due to gas calculation differences. |
```

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/01.deployment-execution.md:
```diff
@@ -0,0 +1,65 @@
+---
+title: EVM bytecode deployment and execution
```
This page feels rather deeply technical

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md:
```diff
@@ -0,0 +1,57 @@
+description:
+---
+::callout{icon="i-heroicons-exclamation-triangle" color="amber"}
+Given that the EVM Bytecode Emulator runs on top of EraVM, the same limitations and differences detailed in
```
This is misleading, only part of limitations is inherited

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md:
```diff
@@ -0,0 +1,57 @@
+  - Only contracts accessed within the EVM execution context are considered warm.
+  - Access from EraVM does **not** affect EVM warm/cold storage tracking.
+- **Contract storage is not cleared on deployment.** Unlike Ethereum, storage in newly created contracts is not explicitly reset.
+- **EVM gas refund mechanisms are not implemented.**
```
```suggestion
- **EVM gas refund mechanisms are not implemented. We use our own gas refunds in EraVM.**
```

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md:
```diff
@@ -0,0 +1,57 @@
+- **Contract storage is not cleared on deployment.** Unlike Ethereum, storage in newly created contracts is not explicitly reset.
+- **EVM gas refund mechanisms are not implemented.**
+- **Intrinsic gas costs differ from Ethereum.**
+- **Contract deployment failures increment nonces.** This matches Ethereum behavior but differs from EraVM,
```
```suggestion
- **Failed deployment of a contract by another contract increments nonce.** This matches Ethereum behavior but differs from EraVM,
```

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md:
```diff
@@ -0,0 +1,57 @@
+
+## Gas Limit Behavior
+
+- The **`GASLIMIT` opcode returns EraVM batch limits in ergs**, not the Ethereum gas limit.
```
Let's say it returns meaningless value

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md:
```diff
@@ -0,0 +1,57 @@
+## Gas Limit Behavior
+
+- The **`GASLIMIT` opcode returns EraVM batch limits in ergs**, not the Ethereum gas limit.
+- **Gas costs differ across opcodes.** Since EVM execution is emulated, each EVM instruction
```
This is confusing. **EVM** gas costs for opcodes are the same as in EVM

uF4No replied:
Good catch. I'll remove the second bullet point, given that this section is about gas limit and that we mention the the higher costs of emulation in other places.

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md:
```diff
@@ -0,0 +1,57 @@
+Some opcodes like CALLCODE, SELFDESTRUCT, BLOBHASH and BLOBBASEFEE are not supported
+on the EVM bytecode emulator as they're not supported on EraVM.
+
+Check [the list of unsupported opcodes](../../differences/evm-instructions)
```
I'd rather list unsupported opcodes here, without this link. That [page](https://zksync-docs-staging-5eb09--pr333-feat-emulator-hcwmu4nw.web.app/zksync-protocol/differences/evm-instructions) is overloaded with info which is not really relevant for the emulator

uF4No replied:
Ok, given that we link to the EVM instructions page in the callout at the start of the page, I think it's ok to just list the 4 unsupported ones.

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/03.evm-differences.md:
```diff
@@ -0,0 +1,57 @@
+Some precompiles like `RIPEMD-160` and `blake2f` are not supported on the EVM bytecode
+emulator as they're not supported on EraVM.
+
+Check [the list of available precompiles](../../differences/pre-compiles)
```
We need to add Identity precompile to the list of available precompiles

uF4No replied:
added

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/05.limitations.md:
```diff
@@ -0,0 +1,88 @@
+---
+title: Limitations and Known Issues
```
This page basically duplicates "Differences Between the EVM Emulator and Ethereum" page. I propose to remove it (probably it was created by AI, page is a bit confusing).

uF4No replied:
yeah, there's some overlap like the precompiles and opcodes but other things are only listed in this page, like access lists, refunds, etc. But as you mentioned above, if we have a specific page for all gas related feaures/limitations/differences, I agree we can delete this page.

0xVolosnikov commented on content/20.zksync-protocol/20.contracts/31.evm-emulator/00.overview.md:
```diff
@@ -0,0 +1,65 @@
+---
```
I think that EVM Eulator section should be in `zksync-protocol` section, not in `zksync-protocol/contracts`

uF4No replied:
ok. It'll be easier to find 🤙

uF4No commented:
> I think we need to add some notification that EVM emulator is an **optional** feature and is not yet available on Era

That's why I've separated the content between the ZKsync Era and ZKsync Protocol sections. This will be merged only after Emulator is on Era testnet. But yeah, we can mention that this is a feature that can be enabled in the Overview page of the ZKsync Protocol section

uF4No committed:
feat: adds identity precompile

uF4No committed:
fix: pr comments

uF4No committed:
Apply suggestions from code review

Co-authored-by: Vladislav Volosnikov <vladyan18@gmail.com>

arose00 commented:
In the `ZKSync Protocol` section we have the `EraVM` section of the docs - did you consider including a mention of the interpreter there? It may be useful to have something like this, as I think the `Unique features` section of the docs is a little buried & if I was an external developer looking for info on this, I'd likely be digging into the Protocol docs first. 

_Editing to add:_ I see this now in the `Contracts` section of the docs. This makes sense, although perhaps discoverability still has some room for improvement so will leave this here in case it sparks an idea.

uF4No committed:
fix: ci job

uF4No committed:
feat: internal links

uF4No commented:
@arose00 We've moved it to the root level inside the "ZKsync Protocol" section but agree we can improve discoverability. I'll find a few places to add internal links.

uF4No committed:
fix: broken links reorg

uF4No committed:
fix: broken link compiler

0xVolosnikov committed:
chore: Add EVM gas emulation page and cleanup (#346)

<!--

Thank you for contributing to the ZKsync Docs!

Before submitting the PR, please make sure you do the following:

- Update your PR title to follow [Conventional
Commits](https://www.conventionalcommits.org/en/v1.0.0/)
- Read the [Contributing
Guide](https://github.com/matter-labs/zksync-docs/blob/main/CONTRIBUTING.md).
- Understand our [Code of
Conduct](https://github.com/matter-labs/zksync-docs/blob/main/CODE_OF_CONDUCT.md)
- Please delete any unused parts of the template when submitting your PR

-->

# Description

<!-- Please describe what are the changes and what they are solving for
in this PR. -->

## Linked Issues

<!-- If you have any issues this PR is related to, link them here. -->
<!--
Check out
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
on how to automate linking a GitHub Issue to a PR.
-->

## Additional context

0xVolosnikov committed:
Fix spelling issues

uF4No committed:
fix: linting issues and typos

uF4No committed:
Merge branch 'main' into feat/emulator

0xVolosnikov commented on content/20.zksync-protocol/75.evm-emulator/04.technical-details.md:
```diff
@@ -24,24 +25,29 @@ When an EVM contract is called, the following steps occur:
-3. **External Calls**
-   - If external call to another EVM contract is requested, EVM emulator pushes new frame in EvmGasManager and invokes new instance of the EVM emulator for that contract.
-   - If external call to EraVM contract is requested, EVM emulator calls that contract directly (EraVM gas is derived from passed EVM gas). When that call returns or fails, EVM emulator continues execution normally. Delegatecall to EraVM contracts is forbidden.
+1. **External Calls**
```
Looks like numeration is broken

uF4No replied:
actually no. We use `1.` for all but Nuxt takes care of showing the correct one 😉

uF4No committed:
Merge branch 'main' into feat/emulator

0xVolosnikov committed:
Add disclaimer about test blockchains

uF4No committed:
feat: adds q in emulator faq

uF4No committed:
feat: internal link era overview

0xVolosnikov committed:
Clarify comment about L1L2 a bit
